File: leetcode/leetcode_cs/maximum-product-subarray.cs
// https://leetcode.com/problems/maximum-product-subarray/

public class Solution {
    public int MaxProduct(int[] nums) {
        int preA,preB,res;
        preA = preB = res = nums[0];
        for (int i=1; i<nums.Length; ++i) {
            var n = nums[i];
            var a = n*preA;
            var b = n*preB;
            preA = Math.Min(n,Math.Min(a,b));
            preB = Math.Max(n,Math.Max(a,b));
            res = Math.Max(res,Math.Max(preA,preB));
        }
        return res;
    }
}

File: leetcode/leetcode_python/maximum-product-subarray.py
from typing import List

"""
https://leetcode.com/problems/maximum-product-subarray/
"""

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        preA = preB = res = nums[0]
        for i in range(1, len(nums)):
            n = nums[i]
            a, b = n * preA, n * preB
            preA, preB = min(n,a,b), max(n,a,b)
            res = max(res, preA, preB)
        return res
        